plugin/core/windows.py

```diff
@@ -587,6 +587,10 @@ def _show_progress(self, params: Dict[str, Any]) -> None:
                 if status_msg:
                     self._sublime.status_message(status_msg)
             elif value.get('kind') == 'end':
+                if value.get('message'):
+                    status_msg = self._progress_string(token, value)
+                    if status_msg:
+                        self._sublime.status_message(status_msg)
                 self._progress.pop(token, None)  # erase stored title and message for token
```

Comment:
My thinking is:
 - If it's a token provided through `initialize` then it should live forever
 - If it's a token provided through `window/workDoneProgress/create` then it also should forever (as it's not bound to any specific message)
 - If it's a token bound to specific request then should be removed when message is handled.

Reply:
Does Julia or Rust re-create token every time they want to report progress?

I suppose it make sense since you might want to have two different progress states for two long-standing operations. Then It would probably make sense to remove on `end` (for the `window/workDoneProgress/create` case).

Reply:
Spec says:

> In case **the client sets up a progress token in the initialize params** (e.g. property workDoneToken) the server is **also allowed** to use that token (and only that token) using the $/progress notification sent from the server to the client.

Meaning that, along with `window/logMessage`, `window/showMessage` and `window/showMessageRequest`, it is also allowed to notify `$/progress` during the initialize phase.

> If it's a token provided through initialize then it should live forever

No, we initiate the allowance of `$/progress` during the special `initialize` phase.

> If it's a token provided through window/workDoneProgress/create then it also should live forever (as it's not bound to any specific message)

No, the server initiates this and ends it with an `end` kind.

IMO this diff is correct. We don't put a workDoneToken in the `initialize` params so server is not allowed to notify `$/progress` during initialize.

Reply:
Afaik Julia uses progress notification only for background package indexing, which probably is initiated only once after the server starts. I agree with all your statements here. However, in case we want to preserve the tokens from `window/workDoneProgress/create` even after `end` notifications, we should at least erase the stored title and message, so that they don't get reused for new notifications which could omit `"message"` (the spec suggests to reuse the last known `"message"` for `report` notifications if omitted).

Reply:
There seems to be some confusion around requests from editor -> server and server initiated progress.

What is implemented here is *server initiated progress*.

A whole other can of worms is *editor initiated progress*.

Reply:
The `initialize` request with its workDoneProgress token is not special. You can put a workDoneProgress token in any request and the server can report progress on that request.

Reply:
I stand corrected.

But IMO we still should reject all progress notifications that don't match registered token.

Reply:
> But IMO we still should reject all progress notifications that don't match registered token.

Agreed.

Reply:
> But IMO we still should reject all progress notifications that don't match registered token.

OK, I will update this PR later today. But we cannot use `pop()` then on `end` notifications, correct? How would it have to be implemented then, considering @rwols assertion that `len(self._progress)` will grow to infinity with my first implementation?

Reply:
> > If it's a token provided through window/workDoneProgress/create then it also should live forever (as it's not bound to any specific message)
> 
> No, the server initiates this and ends it with an `end` kind.

I think we've agreed that we should pop the token on `end`

Reply:
> But we cannot use pop() then on end notifications, correct?

No it's fine as it is. The entire diff is correct in terms of understanding the spec.

Reply:
Alright, that assumes that all `begin` notifications must be preceded by their own `workDoneProgress/create` request first (or otherwise use a token that we could send as `workDoneToken` parameter in requests). That makes sense, but it was not super clear for me from the specs. Let's hope all server implementations assume the same, instead of sending the request only once and reusing that token for potential multiple progresses.

The PR is updated now to ignore unknown tokens.

Reply:
It's the only correct way to interpret the spec because if servers were allowed to both re-use tokens and create new tokens, there would be no way for an editor to know when to delete data associated with tokens, and thus memory usage would grow indefinitely.